core/src/main/java/tech/kwik/flupke/server/impl/Http3ServerConnectionImpl.java

```diff
@@ -316,22 +295,28 @@ else if (frame instanceof DataFrame) {
         return receivedFrames;
     }
 
-    void handleHttpRequest(HeadersFrame headersFrame, QuicStream quicStream, Encoder qpackEncoder) throws ConnectionError {
+    void handleHttpRequest(HeadersFrame headersFrame, QuicStream quicStream) throws ConnectionError {
         String method = headersFrame.getPseudoHeader(HeadersFrame.PSEUDO_HEADER_METHOD);
         String path = headersFrame.getPseudoHeader(HeadersFrame.PSEUDO_HEADER_PATH);
         String auth = headersFrame.getPseudoHeader(HeadersFrame.PSEUDO_HEADER_AUTHORITY);
         DataFramesReader dataFramesReader = new DataFramesReader(quicStream.getInputStream(), maxDataSize);
         HttpServerRequest request = new HttpServerRequestImpl(method, path, auth, headersFrame.headers(), clientAddress, dataFramesReader.getDataFramesStream());
-        HttpServerResponse response = new HttpServerResponseImpl(quicStream, qpackEncoder);
-
+        HttpServerResponse response = new HttpServerResponseImpl(quicStream, encoder);
+        var extendedConnect = method.equals("CONNECT") && headersFrame.getPseudoHeader(HeadersFrame.PSEUDO_HEADER_PROTOCOL) != null;
         try {
             requestHandler.handleRequest(request, response);
             dataFramesReader.checkForConnectionError();
-            dataFramesReader.close();
-            if (!response.isStatusSet()) {
-                response.setStatus(500);
+            boolean noStatus = !response.isStatusSet();
+            if (!extendedConnect && noStatus) {
+                response.setStatus(method.equals("CONNECT") ? 501 : 500);
+            }
+
+            if (extendedConnect && (noStatus || response.status() >= 200 && response.status() < 300)) {
```

Comment:
so you decided that if no status is set by the handler this is ok? I'm not sure whether I agree, but from backwards compatibility point of view this would be ok. At least I would expect a comment to explain.

Reply:
Yeah I think I'll change it